fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/cursors/SortCursorTests.java

```diff
@@ -0,0 +1,398 @@
+            openSimpleRecordStore(context);
+
+            serializer = new SortedRecordSerializer<>(DynamicMessageRecordSerializer.instance(), recordStore.getRecordMetaData(), recordStore.getTimer());
+
+            for (int i = 0; i < 100; i++) {
+                int num = random.nextInt();
+                sortedNums.add(num);
+
+                TestRecords1Proto.MySimpleRecord.Builder record = TestRecords1Proto.MySimpleRecord.newBuilder();
+                record.setRecNo(i);
+                record.setNumValue2(num);
+                recordStore.saveRecord(record.build());
+            }
+            commit(context);
+        }
+        Collections.sort(sortedNums);
+    }
+
+    abstract class MemoryAdapterBase implements MemorySortAdapter<Tuple, FDBStoredRecord<Message>> {
+        @Override
+        public int compare(Tuple o1, Tuple o2) {
+            return o1.compareTo(o2);
+        }
+
+        @Nonnull
+        @Override
+        public Tuple generateKey(FDBStoredRecord<Message> record) {
+            return num2Field.evaluateSingleton(record).toTuple();
+        }
+
+        @Nonnull
+        @Override
+        public byte[] serializeKey(final Tuple key) {
+            return key.pack();
+        }
+
+        @Override
+        public boolean isSerializedOrderReversed() {
+            return false;
+        }
+
+        @Nonnull
+        @Override
+        public Tuple deserializeKey(@Nonnull final byte[] key) {
+            return Tuple.fromBytes(key);
+        }
+
+        @Nonnull
+        @Override
+        public byte[] serializeValue(final FDBStoredRecord<Message> record) {
+            return serializer.serialize(record);
+        }
+
+        @Nonnull
+        @Override
+        public FDBStoredRecord<Message> deserializeValue(@Nonnull final byte[] bytes) {
+            return serializer.deserialize(bytes).getStoredRecord();
+        }
+
+        @Nonnull
+        @Override
+        public MemorySorter.SizeLimitMode getSizeLimitMode() {
+            return MemorySorter.SizeLimitMode.DISCARD;
+        }
+    }
+
+    @Test
+    public void memorySort() throws Exception {
+        final Function<byte[], RecordCursor<FDBStoredRecord<Message>>> scanRecords =
+                continuation -> recordStore.scanRecords(null, null, EndpointType.TREE_START, EndpointType.TREE_END, continuation, ScanProperties.FORWARD_SCAN);
+        final MemoryAdapterBase adapter = new MemoryAdapterBase() {
+            @Override
+            public int getMaxMapSize() {
+                return 20;
+            }
+        };
+
+        List<Integer> resultNums;
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context);
+            try (RecordCursor<FDBStoredRecord<Message>> cursor = MemorySortCursor.create(adapter, scanRecords, timer, null)) {
+                resultNums = cursor.map(r -> TestRecords1Proto.MySimpleRecord.newBuilder().mergeFrom(r.getRecord()).getNumValue2()).asList().get();
```

Comment:
If you were so inclined (entirely up to you), there is now an `asList()` that allows you to also get the final `RecordCursorResult` that terminated the list, so you can assert on `NoNextReason`: https://github.com/FoundationDB/fdb-record-layer/blob/main/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordCursor.java#L296